features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,143 @@
+# Objective
+
+- Provide cached result to Chat Container API calls
+- Compatible to old API interface
+
+# Overview
+
+Container contains a local store for caching resources. The local store caches resources by resource id.
+
+### Store Update
+
+When SDK recieves result from server, for each record, the newly fetched record would always replace the old one with the same record id if there is one.
+
+### API
+
+For fetch API of each resource, there would be a new callback that should return result from cache immediately.
+
+The interface for each language may be different.
+
+```
+getResource(params, callback, cachedCallback?)
+```
+
+The cachedCallback should be optional, thus compatible to the old API interace.
+
+After client app receives result from both callback, client app need to resolve a single result for display. Skygear Chat Container would provide some utility functions to help client app resolve two callback result.
+
+### Resolve result
+
+There are two scenarios, and solutions would be provided by SDK. If client app requires custom resolve logic, they need to implement themselves.
+
+#### 1
+
+```
+---
+ ^
+ | older cached result
+ | ---
+ v  ^
+--- |
+    | newly fetched result
+    v
+   ---
+```
+
+Solution:
+
+- Newly fetched result is appended to the older cached result
+- For overlapped records, the newer one is taken
+
+#### 2
+
+```
+---
+ ^
+ | older cached result
+ v
+---
+(a gap bewtween two sets of result)
+---
+ ^
+ | newly fetched result
+ v
+---
+```
+
+Solution:
+
+- Older cached result is removed, only the newly fetched result is kept
+
+# Changes Required
+
+### Server
+
+- Add a per conversation sequence to `message` table, to determine distance between two given messages.
```

Comment:
Think about the implementation, I guess it is actually better to keep the auto increment ID global, and use a link list like structure (so each message row have a column called prev-message-id)

Advantage:
- server side implementation is easier. I had write a per column auto increment ID before, it is actually quite tricky to be fast, race condition free, and atomic.
- client side will be a bit more complex and slower, but should be minimal (checking if a number missing in a sequence vs checking if a prev-message-id is missing in sequence are also O(n))
- it is future proof, if one day we have a client got a lot of message and we need to have a distributed lock for ID, we don't need to rewrite this logic. Just make sure we use 64bit int as the column, and use a distributed mono increment sequence